_posts/bashar_allabadi/helm-charts-best-practices.md

```diff
@@ -0,0 +1,245 @@
+
+So in summary their operations are different from stateless services.
+
+### Passing Application configurations
+To pass application configuration, we use Environment variables as recommended by the 12 factor application principles (https://12factor.net/) 
+there are different approaches to do this with helm:
+- You can add your env vars as an array in the deployment file itself 
+```
+    spec:
+      serviceAccountName: {{ include "ego.fullname" . }}
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+            - name: SERVER_PORT
+              value: "8081"
+```
+
+- Or you can create a generic non hardcoded way to render your env vars:
+```
+    spec:
+      serviceAccountName: {{ include "ego.fullname" . }}
+      containers:
+        - name: {{ .Chart.Name }}
+          image: "{{ .Values.image.repository }}:{{ .Values.image.tag }}"
+          imagePullPolicy: {{ .Values.image.pullPolicy }}
+          env:
+{{- if .Values.extraEnv }}
+  {{ range $key, $value := .Values.extraEnv }}
+            - name: {{ $key }}
+              value: {{ $value | quote }}
+  {{- end }}
+{{- end }}
+
+```
+
+My experience is that the 2nd approach is easier to maintain and has less overhead because env vars are added and removed often and in the second way there is no need to change anything in the chart itself, it is dynamic enough do that.
+
+### Chart per service vs Common chart
+We started creating a chart per service but after sometime it was clear that for the majority of microservices they all looked the same, regardless of the technology of the service, the charts looked the same. 
+
+So I introduced a generic chart called Stateless-svc (https://github.com/icgc-argo/charts/tree/master/stateless-svc) and it allows users to configure and customize it through providing the values file. This allows us to:
+
+- enforce common practices like labeling, security, adding extra common resources without having to go over all charts
+- maintain only 1 chart for many microservices. no need to create new chart for every new service which cuts the time to get up and running.
+- easier to automate and build processes around since there are no special cases.
+- consistency in configuring charts and estabilishing conventions.
+
+Some charts are more complicated and may not fit in a generic chart, However, for the majority of services, it will save you alot of time.
+
+
+### Secrets Management
+At an early stage of the project, ARGO, we had to decide on how to manage secrets, K8s does provide a `Secret`  resource type, however we decided to go with vault because it basically provides a much rich solution when it comes to secrets management and storage like policies, and different storage backends, etc.
```

Comment:
Is Vault a k8s technology, or is it provided by something else?

Reply:
No, it's a standalone product by Hashicorp https://www.hashicorp.com/products/vault